### DISCUSSION TITLE
some questions regarding creating a reactant solid solution and serpentinization in Reaktoro

### DISCUSSION THREAD

**Discussion by ganlu888** [reaktoro/reaktoro#390] [Q&A]
### Description

Hi, all!

I have some questions regarding creating a reactant solid solution in Reaktoro:

How do you create the reactant solid solution in Reaktoro?

The reactant rock I want to create is composed of:

- 180 wt% olivine (Mg# = 0.90; equivalent to 5.419 moles Mg1.8Fe0.2SiO4 per kg of rock)

- 15 wt% orthopyroxene (Mg# = 0.85; 1.427 moles Mg0.85Fe0.15SiO3 per kg of rock)

- 5 wt% clinopyroxene (Mg# = 0.90; 0.228 moles CaMg0.9Fe0.1Si2O6 per kg of rock)

Assume ideal molecular mixing between the solid solution end members. For these three reactants, should I define the mineral phases as in the example, or do I need to extend the three new species in the SUPCRT98 database? Does chemical equilibrium in Reaktoro include serpentinization?

For orthopyroxene and clinopyroxene, all the information I have is in this picture:
<img width="530" alt="image" src="https://github.com/reaktoro/reaktoro/assets/172543587/ce5bf46f-6540-4518-82cc-4c2eebd3ae75">

### Example

```py
# Define the mineral phases
a0, a1, a2 =0, 0, 0   # the Redlich-Kister parameters (demonstration values!)

Olivine = SolidPhase("Fayalite Forsterite")
Olivine.setActivityModel(ActivityModelRedlichKister(a0, a1, a2))

Orthopyroxene = SolidPhase("Enstatite Ferrosilite")
Orthopyroxene.setActivityModel(ActivityModelRedlichKister(a0, a1, a2))

Clinopyroxene = SolidPhase("Diopside Hedenbergite")
Clinopyroxene.setActivityModel(ActivityModelRedlichKister(a0, a1, a2))

system = ChemicalSystem(db, solution, Olivine, Orthopyroxene, Clinopyroxene)
```

### What is your operating system?

Windows

**Comment by allanleal** [comment]
Hi @ganlu888 - you could create an initial chemical state setting the masses of these minerals and then scaling the masses of the phases as shown below:

```py
# Set molar composition of solid phase Olivine
state.set("Fayalite", 0.123, "mol")
state.set("Forsterite", 0.987, "mol")

# Set molar composition of solid phase Orthopyroxene
state.set("Enstatite", 0.123, "mol")
state.set("Ferrosilite", 0.987, "mol")

# Set molar composition of solid phase Clinopyroxene
state.set("Diopside", 0.123, "mol")
state.set("Hedenbergite", 0.987, "mol")

# Set initial mass percentages of solid phases
state.scalePhaseMass("Olivine", 180, "g")
state.scalePhaseMass("Orthopyroxene", 15, "g")
state.scalePhaseMass("Clinopyroxene", 5, "g")

# Set initial mass percentage of aqueous phase to complete 100 wt%
state.scalePhaseMass("AqueousPhase", 800, "g")

```

**Reply by ganlu888** [reply]
Hi Allan,

Sorry to bother you again. I have a question about how to print the products after reaction equilibrium. I tried using print(state), but the state only contains some ions. I did not see the substances that should be produced by serpentinization, such as serpentine and magnetite. Additionally, the process should produce hydrogen gas at temperatures above 300-350°C and then decrease, but in my results, the hydrogen production keeps increasing. Here is my code. Could you please take a look and identify what might be wrong?

import numpy as np
import matplotlib.pyplot as plt
from reaktoro import *

# Loading thermodynamic databases
db = SupcrtDatabase("supcrt98")

# Define the aqueous phase and the chemical system
solution = AqueousPhase(speciate("Mg Si Ca Na+ Cl- Fe O H+ K+ HCO3- SiO2(aq)"))

# Temperature and pressure values
temperatures = np.array([25.0000, 50.0000, 75.0000, 100.0000, 175.0000, 250.0000, 325.0000, 400.0000])
pressure = 35

# Debye-Huckel parameters
adh_values = np.array([0.5021, 0.5254, 0.5533, 0.5859, 0.7125, 0.9000, 1.2363, 2.5708])
bdh_values = np.array([0.3284, 0.3324, 0.3367, 0.3414, 0.3566, 0.3734, 0.3947, 0.4399])
bdot_values = np.array([0.0410, 0.0435, 0.0450, 0.0460, 0.0480, 0.0335, 0.0000, 0.0000])

# Define the mineral phases
a0, a1, a2 = 0, 0, 0  # the Redlich-Kister parameters (demonstration values!)

Olivine = SolidPhase("Fayalite Forsterite")
Olivine.setName("Olivine")
Olivine.setActivityModel(ActivityModelRedlichKister(a0, a1, a2))

Orthopyroxene = SolidPhase("Enstatite Ferrosilite")
Orthopyroxene.setName("Orthopyroxene")
Orthopyroxene.setActivityModel(ActivityModelRedlichKister(a0, a1, a2))

Clinopyroxene = SolidPhase("Diopside Hedenbergite")
Clinopyroxene.setName("Clinopyroxene")
Clinopyroxene.setActivityModel(ActivityModelRedlichKister(a0, a1, a2))

system = ChemicalSystem(db, solution, Olivine, Orthopyroxene, Clinopyroxene)

# Initialize lists to store results
delta_entropies = []
delta_internal_energies = []
delta_enthalpies = []
delta_gibbs_energies = []
h2_species_amounts = []
ph_values = []

# Loop through temperatures
for temp, adh, bdh, bdot in zip(temperatures, adh_values, bdh_values, bdot_values):
    # Set the Debye-Huckel parameters for the current temperature
    params = ActivityModelDebyeHuckelParams()
    params.aiondefault = adh
    params.biondefault = bdh
    params.bneutraldefault = bdot
    solution.set(ActivityModelDebyeHuckel(params))

    # Create the initial state for each temperature
    state = ChemicalState(system)
    state.setTemperature(temp, "celsius")
    state.setPressure(pressure, "MPa")
    state.set("H2O(aq)", 1, "kg")
    state.set("H+", 0.00000001549, "mol")
    state.set("Na+", 0.464, "mol")
    state.set("Cl-", 0.546, "mol")
    state.set("HCO3-", 0.0023, "mol")
    state.set("Ca+2", 0.0102, "mol")
    state.set("Mg+2", 0.0248, "mol")
    state.set("K+", 0.00098, "mol")
    state.set("SiO2(aq)", 0.00016, "mol")

    # Set molar composition of solid phases
    state.set("Fayalite", 4.877, "mol")
    state.set("Forsterite", 0.542, "mol")
    state.set("Enstatite", 1.213, "mol")
    state.set("Ferrosilite", 0.214, "mol")
    state.set("Diopside", 0.205, "mol")
    state.set("Hedenbergite", 0.023, "mol")

    # Scale initial mass percentages of solid phases
    state.scalePhaseMass("Olivine", 0.8, "kg")
    state.scalePhaseMass("Orthopyroxene", 0.15, "kg")
    state.scalePhaseMass("Clinopyroxene", 0.5, "kg")


    # Calculate initial properties
    initial_props = ChemicalProps(state)

    # Solve equilibrium for the given temperature
    solver = EquilibriumSolver(system)
    solver.solve(state)

    # Calculate final properties
    final_props = ChemicalProps(state)
    final_aprops = AqueousProps(state)

    # Compute deltas
    delta_entropy = final_props.entropy() - initial_props.entropy()
    delta_internal_energy = final_props.internalEnergy() - initial_props.internalEnergy()
    delta_enthalpy = final_props.enthalpy() - initial_props.enthalpy()
    delta_gibbs_energy = final_props.gibbsEnergy() - initial_props.gibbsEnergy()

    # Store the results
    delta_entropies.append(delta_entropy)
    delta_internal_energies.append(delta_internal_energy)
    delta_enthalpies.append(delta_enthalpy)
    delta_gibbs_energies.append(delta_gibbs_energy)

    # Store the species amounts
    h2_species_amounts.append(final_props.speciesAmount("H2(aq)"))

    # Calculate and store pH
    ph_values.append(final_aprops.pH())

# Plot the results for thermodynamic properties
plt.figure(figsize=(14, 10))

plt.subplot(3, 2, 1)
plt.plot(temperatures, delta_entropies, label='Δ Entropy', color='blue')
plt.xlabel('Temperature (°C)')
plt.ylabel('Δ Entropy (J/K)')
plt.title('Δ Entropy vs Temperature')
plt.legend()

plt.subplot(3, 2, 2)
plt.plot(temperatures, delta_internal_energies, label='Δ Internal Energy', color='green')
plt.xlabel('Temperature (°C)')
plt.ylabel('Δ Internal Energy (J)')
plt.title('Δ Internal Energy vs Temperature')
plt.legend()

plt.subplot(3, 2, 3)
plt.plot(temperatures, delta_enthalpies, label='Δ Enthalpy', color='red')
plt.xlabel('Temperature (°C)')
plt.ylabel('Δ Enthalpy (J)')
plt.title('Δ Enthalpy vs Temperature')
plt.legend()

plt.subplot(3, 2, 4)
plt.plot(temperatures, delta_gibbs_energies, label='Δ Gibbs Energy', color='purple')
plt.xlabel('Temperature (°C)')
plt.ylabel('Δ Gibbs Energy (J)')
plt.title('Δ Gibbs Energy vs Temperature')
plt.legend()

plt.subplot(3, 2, 5)
plt.plot(temperatures, ph_values, label='pH', color='orange')
plt.xlabel('Temperature (°C)')
plt.ylabel('pH')
plt.title('pH vs Temperature')
plt.legend()

plt.tight_layout()
plt.show()

# Plot the results for species amounts
plt.figure(figsize=(14, 8))

plt.subplot(3, 1, 1)
plt.plot(temperatures, h2_species_amounts, label='H2(aq) Species Amount', color='blue')
plt.xlabel('Temperature (°C)')
plt.ylabel('H2 Species Amount (mol)')
plt.title('H2 Species Amount vs Temperature')
plt.legend()

plt.tight_layout()
plt.show()

**Reply by allanleal** [reply]
Hi @ganlu888 , I have been on vacation since then. I'll ping @gdmiron who might be familiar with your system. You may also consider using ThermoFun databases for this problem.

**Reply by ganlu888** [reply]
Hi, allan @allanleal , Do you know how to output the mass or volume of the solids generated after a reaction reaches equilibrium? I tried using print(final_props) and print(final_aprops), but I could only get information on the quantity or concentration of ions or elements, not the specific mass of solids. For example, I want to know the amount of magnetite produced during serpentinization at a specific temperature. When I checked aprops, it only provided the saturation ratio and saturation index for specific substances. Is there a command that can yield the quantity of a particular species? 
<img width="473" alt="image" src="https://github.com/user-attachments/assets/3e646699-58f0-4cb4-8ef3-4a060f23d348">
<img width="458" alt="image" src="https://github.com/user-attachments/assets/e08871ca-e459-413c-b9de-8ba9eb97844f">

**Reply by allanleal** [reply]
I was out of office. You could use:

~~~py
props.phaseProps(mineralphase).mass()
props.phaseProps(mineralphase).volume()
~~~

where `props` is an object of `ChemicalProps` and `mineralphase` is the name of the mineral (e.g., `"Diopside"`).
